### DISCUSSION TITLE
Imported group region information is not imported

### DISCUSSION THREAD

**Discussion by OSAwebsite** [code4recovery/12-step-meeting-list#1145] [Q&A]
### Contact Details

os.website.a86@gmail.com

### Website With Issue

al-anon.alateen.on.ca

### What happened?

I put this in as a bug, because I feel this is unexpected behaviour, or at least, should be described in the Import information section of the tool.

Here is what happened:
I made a change to a meeting, that impacted information present in the group section of the website listing of meetings.
The impact was on the list of regions for a particular district. To save myself the trouble of having to go through the list of meetings and correct the regions where meetings are held, I first deleted all meeting for that district.
I then performed an import, with no deleting of listings. The import csv showed the new list of regions for all meetings in that district. However, the region  list was not updated correctly after re-importing all meetings of this district.

I experimented with this unexpected result and found out that the import will ONLY work if I first remove the incorrect region from the region table, add the new region to it and then perform an import with no delete (I had, after all, deleted all meetings for the affected district before performing the import).

What is the best way to update this kind of information. Should I have corrected the regions first, or should I have chose the option to delete the meetings that are present in the csv? Or is it reasonable to expect the import to add new regions automatically? And accept the information in the import, as it has been set out?

I hope this makes sense. I would expect that, at all times, an import is reflected in the resulting web listings. That is, after all, why we perform an import.

I now feel insecure that the list of regions per district is correct for my entire list of meetings. Which means, I will be inclined to perform a re-import of the entire list of over 300 meetings in my area.

Thanks for considering this topic.

Nicky
Yours in service

### What browsers are you seeing the problem on?

Chrome, Safari, Microsoft Edge

**Comment by anchovie91471** [comment]
Nicky,

I'm having some difficulty understanding the steps you took as described. Would you mind posted some screenshots that highlight or circle the affected areas of your site so I'm sure I'm understanding what happened?

i.e. When you mention the "group section" do you mean the Group/Location column of the meetings table? You mentioned the "list of regions for a particular district", are you referring to the listing on the left side of an individual meeting page?

Also, can you include the CSV file you attempted to import?

Thank you!
Anthony B

**Reply by tim-rohrer** [reply]
@OSAwebsite (Nicky), are you saying you expected the structure of defined regions to be updated by an import of meetings that have region information defined?

**Comment by OSAwebsite** [comment]
Tim: yes, that is correct.
I have found that this is not happening.
So I will from now on first edit regions when I know it has to be changed. Either by deleting a region where groups are no longer meeting, or adding an additional region. This is a bit of a pain in the you know what but when I do that, it goes well in combination with the second import option to delete listings that are in the .csv.
It would be wonderful if this would happen automatically, but I don't see any import option for regions or districts.

From memory, I seem to remember that they are all auto created upon a fresh new import.
But apparently not when I add additional new meetings?

**Reply by tim-rohrer** [reply]
We can look at that, but right off the bat I would be concerned we’d be messing with another user’s work flow. I also don’t have a lot of indications that people change regions that frequently. Hopefully, this is a one time situation.

**Comment by OSAwebsite** [comment]
Well, now that Covid is less prevalent, it happens a lot at our side when temporarily inactive meetings pick back up in person in another region. I think this past month happened 3 times. I don't always instantly check my lists of cities/towns where meetings happen in a particular district, but I have now added some automated warnings when that scenario happens so I don't forget. In general, you are right, that list would change only when the last meeting leaves a town. Or gets disbanded. Our phone line volunteers look at the District Region lists as well for newcomers, to see if there is a meeting in their town, even before they search the website.

My main worry here is not that I want the TSML to automate things. It is that I do not have a good understanding of how the content of Regions and Districts is created.

I seem to remember that on the very first import both tables are auto filled. Is that correct?
If so, then it is reasonable to expect that a brand new meeting in a region where no meetings were held before, is also auto-entered.
If not, then I think it is a bit confusing.

How you arrange things in the backend of the plugin is not my business. You may need the two databases/tables with regions and districts for certain features in the plugin. That is fine. 

But when you allow your users to create an import sheet, then the information they put in it, should be imported as is. 
If the information is altered because you do an interpretation on the information, then I would like to have that documented so that I know what I can expect.

It may not be top of mind for many, but I do remember several members sighing with relief when they found out that it is possible (for me at least) to remove certain meeting regions from their district list, because they had to keep explaining to old and new members that the region had not had a meeting for a long time.

If you can tell me the workflow that works well for this, then I know how to work with my data.

Thanks for your consideration!

**Comment by anchovie91471** [comment]
Hi @OSAwebsite if I delete meetings for a region I see that the region persists in the regions dropdown, which is definitely a bug, we'll make a ticket to fix that.

Are there other issues? If so, could you walk us through the specific steps to reproduce? Thanks

**Comment by OSAwebsite** [comment]
Hi all who are following this topic,

First off, I understand that this is an issue that is hard to follow.
If I can help, I will certainly try to.
So I will try and do step by step testing.

**Test 1**
•	I deleted all meetings from my test site (which has appr the same number of meetings as the live site).
o	I checked the Regions and District databases within the TSML plugin: all data still there.
•	Then I emptied the trash.
o	I checked the Regions and District databases within the TSML plugin: all data still there.
•	I double-check for the meeting that I worked with this afternoon, that triggered this whole discussion. The place that was new for the list of regions is not currently present in the plugin. The place that should replace, is not there either. It is possible that I deleted that myself earlier. I don't remember.
•	I now import my full list of meetings, using the format I always use when importing. 
o	I choose the first option: don't delete anything. Because I have emptied everything in the meeting list.The list comes in without errors.
•	I check the list of districts and regions in the group section of the details page of the listing and all seems to be well. But when I search for  Egmondville, it does not show up.Instead, it shows the name of the place that is defined for the meeting's location: Huron East (Egmondville). So that makes me think that the regions are in fact meeting cities and towns an derived from the address in the listing.

**Test 2**
Now I want to recreate the circumstances of earlier this afternoon. I change the meeting place from Egmondville to Clinton. Just like this afternoon, this causes the list of regions for the district this meeting belongs to change. And instead of going through the website for each of its 12 meetings and manually change the list of regions, I decide to update all meetings with a new import. I do not delete the meetings, I choose the second option during import: delete only the meetings, locations, and groups for the regions present in this CSV

The result is a duplicate meeting in an otherwise full list. The one where I changed the region now has 2 listings. One with Edgmondville and one with Clinton. When I go to the details page, both listings do not show a region name. It just says 'Region'. I have not seen this happen before.
o	I checked the Regions and District databases within the TSML plugin: this now contains both Edgmondville and Huron East (Edgemondville). Clinton is not there. The other D5 meetings show correct region names.

 
**Test 3**
I manually delete the two duplicates. And do a fresh import of just that one meeting. For testing purposes I stick with Clinton as the region name (a.k.a. the meeting location). To see if the database will update when I import with the option 'Delete nothing'.
The meeting result still shows 'region' instead of Clinton. I can see that Clinton does not exist in the Regions database.

**Test 4**
I manually add Clinton to the Regions database. I delete the incorrect listing. I re-import from the same (but renamed) csv importsheet.
The region name is still 'Region' but now I can choose Clinton manually from the pop-up.
I reset the region to Clinton manually.

**Test 5**
Now I do what I did this afternoon, but not for the full set, just for this one meeting. I reset Clinton to Edgmondville. Both of these regions now exist in the Regions database and I expect the region to now being entered properly. I set the import to 'only delete meeting at this region', the second one. The import keeps showing Clinton in the region. I can now choose to set it manually to Edgmondville. 

I now notice that I only have 2 meetings in my list. Somewhere along the line all others were deleted... with only 3 in thrash. I saw this happen twice in this sequence, and always noticed too late when all were deleted.

**Test 6:**
Repeated all the previous testing and the results were as expected: good, with no errors.


The thing I noticed half way through is this:
I made a typo in the name of the place. I typed Edgemondville instead of Edgmondville.
This was just during testing, not today in the live website.

Is it possible that a region that is not recognised by Google can block the update of some of the other information in the Importsheet? If that is the case, it seems to me this should throw a visible error. It seems unlikely to me, because the full address was found without any issues. The Region looks like an additional field of the plugin, rather than something that is linked with Google?

I don't know. I just know that I did these tests 3 times. The first time a lot of errors happened. The second time there was a new one when the region was not displayed and I lost listings without having thrashed them myself. The third time when I wanted to demonstrate this and pinpoint where it happened, everything went as expected. By then, of course, all of the three different regions I used were in the Regions database. So it seems that there is something there that causes the confusion.

I am stopping the testing now... If you need me to do other testing, let me know.

**Reply by tim-rohrer** [reply]
If possible, let's try to keep future comments on **this particular thread**. That will help others follow along.

As @anchovie91471 mentioned, we acknowledge that it might be a bug if deletion of the last meeting in a region doesn't delete the region, but I want to talk with other team members more before we change it. I'm thinking of a scenario when someone is setting up a new site so they add a region, then add a meeting for the region, then realize that they don't have it right, so they delete the meeting and now the region gets deleted as well. They'll be chasing their tail a bit because they'll have to reenter the region again. True, this is another edge case.

We'll go through and try to recreate some of your tests. I do want to try and level expectations by clarifying a couple of points.

1. You wrote elsewhere:
>I seem to remember that on the very first import both tables are auto filled. Is that correct?
>If so, then it is reasonable to expect that a brand new meeting in a region where no meetings were held before, is >also auto-entered.
>If not, then I think it is a bit confusing. 

When I just tested this, as I expected, the new region was automatically added. I imported a CSV containing a single meeting in Region. Then I changed the csv so the meeting was in Nearby Region and then imported it again, using option 1 for the import (nothing deleted). Both regions are now showing in the regions table.

2. The regions listed in their specific table do not come from Google. They are either manually entered or created during the CSV import. Doing it this way is desired because we want this to be flexible.

Let us work with some of your other tests for a bit.

**Reply by tim-rohrer** [reply]
@OSAwebsite, please send us a copy of the csv. Some of what I'm reading in your tests just doesn't make sense to me and I'm hoping that being able to read the csv will help.

Please send to devs@code4recovery.org.

Thanks.
